src/services/ProgressService.ts

```diff
@@ -0,0 +1,142 @@
+import Message from '../models/Message';
+import { MessageStatus } from '../repositories/MessageRepository';
+import Utils from '../Utils';
+import { AUXILIARY_GAS_PRICE, ORIGIN_GAS_PRICE } from '../Constants';
+
+const Mosaic = require('@openst/mosaic.js');
+const web3utils = require('web3-utils');
+
+/**
+ * It facilitates progress staking and minting.
+ */
+class ProgressService {
+  private originWeb3: any;
+
+  private auxiliaryWeb3: any;
+
+  private gatewayAddress: string;
+
+  private originWorkerAddress: string;
+
+  private auxWorkerAddress: string;
+
+  /**
+   * Constructor of ProgressService class.
+   * @param originWeb3 Origin chain web3 object.
+   * @param auxiliaryWeb3 Auxiliary chain web3 object
+   * @param gatewayAddress Address of gateway contract.
+   * @param originWorkerAddress Origin chain worker address.
+   * @param auxWorkerAddress Auxiliary chain worker address.
+   */
+  public constructor(
+    originWeb3: any,
+    auxiliaryWeb3: any,
+    gatewayAddress: string,
+    originWorkerAddress: string,
+    auxWorkerAddress: string,
+  ) {
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.gatewayAddress = gatewayAddress;
+    this.originWorkerAddress = originWorkerAddress;
+    this.auxWorkerAddress = auxWorkerAddress;
+  }
+
+  /**
+   * This method react on changes in Message model.
+   * @param messages List of Message models.
+   */
+  public async update(messages: Message[]): Promise<void> {
```

Comment:
What do you think of this implementation? 

```
  public async update(messages: Message[]): Promise<void> {
    const progressPromises = messages.map(async (message) => {
      if (message.sourceStatus === MessageStatus.Declared
        && message.targetStatus === MessageStatus.Declared
      ) {
        return Promise.all([this.progressStake(message), this.progressMint(message)]);
      }

      if (message.sourceStatus === MessageStatus.Declared) {
        return this.progressMint(message);
      }

      if (message.targetStatus === MessageStatus.Declared) {
        return this.progressMint(message);
      }
      return Promise.resolve();
    });

    await Promise.all(progressPromises);
  }

```

Reply:
Done